### PR TITLE
error handling in backend

### PR DESCRIPTION
**Pull request by Minji1234**
I added error handling feature in oga/backend/views.py file.
Changes are following

login_required decorator for checking whether the user is logged in.
check_request decorator for catch KeyError and ValueError
Applied those decorators to some functions.
Make the users to check their own profile(not others' profile)
Before detail of question check if there is matching question or not

### CODE REVIEW

File: oga/backend/users/views.py [hunk 1]
```diff
@@ -1,42 +1,68 @@
...
+            return HttpResponse(status=400)
+    return wrapper
+
+def check_request(views_func):
+    @wraps(views_func)
+    def wrapper(request, *args, **kwargs):
+        try:
+            return views_func
+        except (KeyError, ValueError) as e:
+            return HttpResponse(status=400)
```
Comment (taehioum) [inline oga/backend/users/views.py]: I think using something like
```python
return HttpResponseBadRequest(str(e))
```
might be better for understanding code.

File: oga/backend/users/views.py [hunk 2]
```diff
@@ -1,42 +1,68 @@
...
 
-def index(request):
-    return HttpResponse('Hello World!')
+# def index(request):
+#     return HttpResponse('Hello World!')
 
+@login_required
+@check_request
 def UserProfile(request, username):
+    if not (request.user.username == username):
```
Comment (taehioum) [inline oga/backend/users/views.py]: Django has something along the lines of check_permission decorator. Maybe we can use them instead.

Also, I think dividing user profile view to two (one for self, and one for viewing others perhaps?)
might be a better idea, design choice wise. Since this commit is not directly related to error_handling, I think we should discuss our design choice before implementing this part.

File: oga/backend/users/views.py [hunk 3]
```diff
@@ -1,42 +1,68 @@
...
 def Main(request, username):
     question_user = get_object_or_404(Profile, username=username)
     question_list = [question for question in Question.objects.filter(author = question_user).values()]
     return render(request, 'users/main.html', {'user': question_user, 'question_list': question_list})
     
 # Displays detailed question page
+@login_required
+@check_request
 def Details(request, question_id):
+    if not (Question.objects.filter(id==question_id).exists()):
```
Comment (taehioum) [inline oga/backend/users/views.py]: This is not necessary as
```python
question_to_answer = get_object_or_404(Question, id = question_id)
```
already handles this.

File: oga/backend/users/views.py [hunk 4]
```diff
@@ -46,33 +72,35 @@ def questions(request):
...
-            return HttpResponseBadRequest()
+        req_data = json.loads(request.body.decode())
+        username = req_data['username']
+        password = req_data['password']
         user = authenticate(username=username, password=password)
         if user is not None:
             login(request, user)
             response_dict = {'id': user.id}
-            return JsonResponse(response_dict, status=201)
+            return JsonResponse(response_dict, status=204)
```
Comment (taehioum) [inline oga/backend/users/views.py]: If we are not going to return anything upon successful login, please remove the response dict.
Otherwise, please change the status=204 to 201

File: oga/backend/users/views.py [hunk 5]
```diff
@@ -46,33 +72,35 @@ def questions(request):
...
         else:
             return JsonResponse({}, status=401)
+    else:
```
Comment (taehioum) [inline oga/backend/users/views.py]: This can be checked by django's 
```python
@require_http_methods(["GET", "PUT", "DELETE"])
```
decorator.
Please remove the ifs and use this.`

File: oga/backend/users/views.py [hunk 6]
```diff
@@ -1,42 +1,68 @@
...
 # Displays detailed question page
+@login_required
+@check_request
 def Details(request, question_id):
+    if not (Question.objects.filter(id==question_id).exists()):
+        return HttpResponse(status=404)
     question_to_answer = get_object_or_404(Question, id = question_id)
     # questions to answers are one to many
     answer_list = [answer for answer in Answer.objects.filter(question = question_to_answer).values()]
     return render(request, 'users/detail.html', {'question': question_to_answer, 'answer_list': answer_list})
```
Comment (taehioum) [inline oga/backend/users/views.py]: Please return a JsonObject instead of render if we are going to use this API in order to get list of answers to a question.

Comment (taehioum) [inline oga/backend/users/views.py]: Please return a JsonObject instead of render if we are going to use this API in order to get list of answers to a question.

File: oga/backend/users/views.py [hunk 7]
```diff
@@ -1,42 +1,68 @@
...
 from django.views import generic
+from functools import wraps
 
+def login_required(views_func):
+    @wraps(views_func)
+    def wrapper(request, *args, **kwargs):
+        if request.user.is_authenticated:
+            return views_func(request, *args, **kwargs)
+        else:
+            return HttpResponse(status=400)
```
[the remark this answers is not in the document]
Comment (Minji1234) [reply oga/backend/users/views.py]: I fixed

File: oga/backend/users/views.py (hunk 1 above)
Comment (Minji1234) [reply oga/backend/users/views.py]: thanks. I changed the code as you said

File: oga/backend/users/views.py (hunk 2 above)
Comment (Minji1234) [reply oga/backend/users/views.py]: You mean from django.contrib.auth.decorators import permission_required?
As you mentioned I searched it in the internet but it looks like it deals with what kind of things can be done regarding specific kind of user group.
In this case, we just need to find if there username matches or not.
If I didn't get your words well, then could you fix me?

File: oga/backend/users/views.py (hunk 3 above)
Comment (Minji1234) [reply oga/backend/users/views.py]: Thanks. I deleted.

File: oga/backend/users/views.py (hunk 4 above)
Comment (Minji1234) [reply oga/backend/users/views.py]: I changed it to status=201

File: oga/backend/users/views.py (hunk 5 above)
Comment (Minji1234) [reply oga/backend/users/views.py]: Thanks. I left if condition for future implementation but removed 'else:'

File: oga/backend/users/views.py (hunk 2 above)
Comment (taehioum) [reply oga/backend/users/views.py]: Thank you for the information. I probably should search more before leaving this kind of comments :(. I think you are right.

File: oga/backend/users/views.py (hunk 1 above)
Comment (taehioum) [inline oga/backend/users/views.py]: I think we don't have to specify request argument here, since we don't use it unlike login_required.

```suggestion
    def inner(*args, **kwargs):
        try:
            return view_func(*args, **kwargs)
        except (KeyError, json.JSONDecodeError) as e:
            return HttpResponseBadRequest(str(e))
```

perhaps something like this would be better..?

Also, as of now, because we only return views_func (instead of ```views_func(request, *args, **kwargs) ```executing this line causes error along the lines of ```AttributeError: 'function' object has no attribute 'get'```

File: oga/backend/users/views.py [hunk 8]
```diff
@@ -1,42 +1,68 @@
...
+    def wrapper(request, *args, **kwargs):
+        try:
+            return views_func
+        except (KeyError, ValueError) as e:
+            return HttpResponse(status=400)
+    return wrapper
 
-def index(request):
-    return HttpResponse('Hello World!')
+# def index(request):
```
Comment (taehioum) [inline oga/backend/users/views.py]: I think removing this part causes error during ```python3 manage.py runserver```, since urls.py has a url path to views.index.
I would be grateful if you could please update urls.py as well.

File: oga/backend/users/views.py (hunk 1 above)
Comment (Minji1234) [reply oga/backend/users/views.py]: Great. I follow your suggestions.

File: oga/backend/users/views.py (hunk 8 above)
Comment (Minji1234) [reply oga/backend/users/views.py]: Okay. I will remove index url in urls.py
